Documentation/gettingstarted/cni-chaining-azure-cni.rst

```diff
@@ -96,6 +96,12 @@ This will create both the main cilium daemonset, as well as the cilium-node-init
 existing Azure CNI plugin to run in 'transparent' mode.
 
 .. include:: k8s-install-restart-pods.rst
-.. include:: k8s-install-validate.rst
+
+Validate the Installation
+=========================
+
+.. include:: cli-download.rst
+.. include:: cli-status.rst
+.. include:: cli-connectivity-test.rst
```

Comment:
These six lines are copy-pasted in a lot of places throughout the documentation. Would we be able to get the best of both worlds, have individual pieces such as `cli-download` etc, but also a single include for the whole thing like `cli-validate-installation`?

That should give us the flexibility to either include the whole thing (for when it inevitably changes or maybe even has more internal includes), or individual pieces.

Reply:
I had this at first, but I am not sure this is good design, hence why I removed it.

As exemplified by the structure we had beforehand, regrouping the title + the three bits in a re-usable component makes it more likely that changes will have side-effects and that refactors are hard to properly pin down. I think that in most cases we should strive to never nest more than 1 indirection when it comes to templating, i.e. including a unit is fine, but including something that itself includes something else is not. Most of the issues I've seen in the docs or docs PR so far come from the difficulty of knowing how the final result will look like when we only deal with the `rst`, and from side-effects due to someone updating a generic block somewhere and not knowing about the context of where it's used (i.e. even if you do a quick grep to find usage, you are likely to miss the context or some other part of the actual template that includes what you update, because it is actually nested in another template etc.).

Hence, I believe having the individual bits to assemble as you please with the titles that you want is how it should be. It just so happens that several guides have the three bits at once with the same title right now, but that is not systematic as you have seen.

@qmonnet Any opinion on that?

Reply:
I definitely understand your side of the argument. I just wanted to have a discussion on the subject. We don't necessarily have to block this PR on it, but it would be good to have a consistent direction moving forward.

To me it seems like `Validate the installation` is the _actual block_ here. The fact that it's made from three under the hood imports sort of like an implementation detail. 

Say we add a fourth block to the validate the installation guide; we then have to go find all the references to `validate the installation` and then update the imports. We generally don't like to copy-paste the same thing in code more than 2-3 times, so I don't see why in docs it's ok to do 12 times.

> Most of the issues I've seen in the docs or docs PR so far come from the difficulty of knowing how the final result will look like when we only deal with the rst

Inability to see the final result while doing the PR seems like a symptom, not the problem itself.

Reply:
> To me it seems like Validate the installation is the actual block here. The fact that it's made from three under the hood imports sort of like an implementation detail.

This is where I disagree: as we can see from the guides, it is only because we decided to swap out the old connectivity test with the new CLI connectivity test that all Helm guides end up needing this specific arrangement. Guides which use the CLI at any point beforehand will set it up in a different way. So, I don't feel like `Validate the installation` is the actual block, it's just a coincidence that Helm guides need `cli-download` (which is unquestionably not "validating the installation") before doing the other two. It's also a coincidence that we always have `cli-status` immediately followed by `cli-connectivity-test`: they are not doing the same thing, one is validating the Cilium agents and operator (this one is unquestionably "validating the installation), the other is checking connectivity (which arguably may or may not be part of "validating the installation").

> Say we add a fourth block to the validate the installation guide; we then have to go find all the references to validate the installation and then update the imports. We generally don't like to copy-paste the same thing in code more than 2-3 times, so I don't see why in docs it's ok to do 12 times.

Same answer: `Validate the installation` is not the block. If we extend the CLI and need to add another `cli-foo` block, then it will have to go through the same process of "which guide needs it where".

Reply:
> So, I don't feel like Validate the installation is the actual block, it's just a coincidence that Helm guides need cli-download (which is unquestionably not "validating the installation") before doing the other two. It's also a coincidence that we always have cli-status immediately followed by cli-connectivity-test: they are not doing the same thing, one is validating the Cilium agents and operator (this one is unquestionably "validating the installation), the other is checking connectivity (which arguably may or may not be part of "validating the installation").

The fact that `validate the installation` block contains first the `download the CLI` block is most certainly an implementation decision, not a requirement. You could just as well move the `download the CLI` to it's own block which may or may not be included above, or even present once at the top of the guide (if it's applicable to the whole thing). You could also add something like a disclaimer saying "cilium-cli is required. If you haven't got it click <here≥". Not all the guides always need to always inline the instructions in their entirety. I think it's ok to click on another page to do XYZ, such as install another binary.

Polymorphing the meaning of `validate the installation` based on the context seems to lead to a lot of this complexity. I don't see how in the long term your approach helps to mitigate that complexity.

Reply:
> You could just as well move the download the CLI to it's own block which may or may not be included above, or even present once at the top of the guide (if it's applicable to the whole thing). 

Yes, and that's exactly what happens in the non-Helm guides 😛

> You could also add something like a disclaimer saying "cilium-cli is required. If you haven't got it click <here≥". Not all the guides always need to always inline the instructions in their entirety. I think it's ok to click on another page to do XYZ, such as install another binary.

Yep, agreed. I mostly tried to keep the existing structure we had as-is, but it is indeed not necessary that all sections be self-contained.

> Polymorphing the meaning of validate the installation based on the context seems to lead to a lot of this complexity. I don't see how in the long term your approach helps to mitigate that complexity.

It's a trade-off: instead of nesting complexity, you have duplication complexity. My gut feeling having worked with "nested flat templating" systems in the past (including `rst`) is that when you have a big group of people collaboratively editing, most people seem to have an easier time with duplication than nesting, because if includes are only 1-deep you can easily drop into and edit any document without prior knowledge (including the building blocks, because it is easy to list all places they are included), whereas with deep nesting it becomes very complex to properly make a mental map of the whole thing and identify the side effects (this is exactly what happened in the PR that lead to the issues that we're trying to address in this PR 😅).

But honestly, I don't think either solution is better, "nested flat templating" is always hard no matter whether it is `rst` or Helm templates... It'd be better if we had generators with smarter templating (e.g. variables which you can pass to the block so that it may or may not include the `Install CLI` bit or the `Validate the installation` header itself), like CUE or Jsonnet are to Json/YAML... but we don't have that for now 😁

If you agree, I think this discussion can be summarized as a choice between:

- NOK for nesting blocks, because the intent is to make it easier to understand at a glance what a template actually means, no matter if there is duplication in similar guides.
- OK for nesting blocks, because the intent is to reduce duplication, no matter if understanding the structure is harder because inclusion context might be hidden away behind several nested includes.

My vote is for NOK, but I have no strong feeling for either (it's just a gut feeling).

Reply:
@glibsm I am moving back the blocks in an intermediate template so that we can have both the CLI and old connectivity tests in tabs, as discussed in the community meeting. We can resume the discussion regarding the general structure in a future community meeting :)